src/txBaseMsgProcessor.cpp

```diff
@@ -0,0 +1,444 @@
+        if (itPrev->second.empty())
+            mapOrphanTransactionsByPrev.erase(itPrev);
+    }
+
+    mapOrphanTransactions.erase(it);
+}
+
+bool TxBaseMsgProcessor::IsOrphan(const uint256& txBaseHash) const
+{
+    LOCK(cs_orphanTxes);
+    return mapOrphanTransactions.count(txBaseHash) != 0;
+}
+
+const CTransactionBase* TxBaseMsgProcessor::PickRandomOrphan()
+{
+    LOCK(cs_orphanTxes);
+    uint256 randomhash = GetRandHash();
+    std::map<uint256, COrphanTx>::iterator it = mapOrphanTransactions.lower_bound(randomhash);
+    if (it == mapOrphanTransactions.end())
+        it = mapOrphanTransactions.begin();
+
+    return (it->second.tx).get();
+}
+
+unsigned int TxBaseMsgProcessor::LimitOrphanTxSize(unsigned int nMaxOrphans)
+{
+    LOCK(cs_orphanTxes);
+    unsigned int nEvicted = 0;
+    while (mapOrphanTransactions.size() > nMaxOrphans)
+    {
+        EraseOrphanTx(PickRandomOrphan()->GetHash());
+        ++nEvicted;
+    }
+    return nEvicted;
+}
+
+unsigned int TxBaseMsgProcessor::countOrphans() const
+{
+    LOCK(cs_orphanTxes);
+    return mapOrphanTransactions.size();
+}
+
+void TxBaseMsgProcessor::EraseOrphansFor(NodeId peer)
+{
+    LOCK(cs_orphanTxes);
+    int nErased = 0;
+    std::map<uint256, COrphanTx>::iterator iter = mapOrphanTransactions.begin();
+    while (iter != mapOrphanTransactions.end())
+    {
+        std::map<uint256, COrphanTx>::iterator maybeErase = iter++; // increment to avoid iterator becoming invalid
+        if (maybeErase->second.fromPeer == peer)
+        {
+            EraseOrphanTx(maybeErase->second.tx->GetHash());
+            ++nErased;
+        }
+    }
+    if (nErased > 0) LogPrint("mempool", "Erased %d orphan tx from peer %d\n", nErased, peer);
+}
+// End of Orphan Txes/Certs Tracker section
+
+// Rejected Txes/Certs Tracker section
+void TxBaseMsgProcessor::SetupRejectionFilter(unsigned int nElements, double nFPRate)
+{
+    LOCK(cs_rejectedTxes);
+    recentRejects.reset(new CRollingBloomFilter(nElements, nFPRate));
+}
+
+bool TxBaseMsgProcessor::HasBeenRejected(const uint256& txBaseHash) const
+{
+    LOCK(cs_rejectedTxes);
+    assert(recentRejects);
+    return recentRejects->contains(txBaseHash);
+}
+
+void TxBaseMsgProcessor::MarkAsRejected(const uint256& txBaseHash)
+{
+    LOCK(cs_rejectedTxes);
+    assert(recentRejects);
+    recentRejects->insert(txBaseHash);
+}
+
+void TxBaseMsgProcessor::RefreshRejected(const uint256& currentTipHash)
+{
+    LOCK(cs_rejectedTxes);
+    assert(recentRejects);
+    if (currentTipHash != hashRecentRejectsChainTip)
+    {
+        // If the chain tip has changed previously rejected transactions
+        // might be now valid, e.g. due to a nLockTime'd tx becoming valid,
+        // or a double-spend. Reset the rejects filter and give those
+        // txs a second chance.
+        hashRecentRejectsChainTip = currentTipHash;
+        recentRejects.reset(nullptr);
+        recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
```

Comment:
It seems like the first reset (with _nullptr_) is not necessary.

```suggestion
        recentRejects.reset(new CRollingBloomFilter(120000, 0.000001));
```